### PR TITLE
Introduce error constructor and corresponding binding patterns

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
This PR align ballerina error constructor syntax with the spec, by fixing https://github.com/ballerina-platform/ballerina-lang/issues/14670 and https://github.com/ballerina-platform/ballerina-lang/issues/15201

## Samples
Error details are provided as named arguments.
```ballerina
error er1 = error("error1", message = "msg1");

type TrxErrorData record {|
    string message = "";
    error? cause = ();
    string data = "";
|};

TrxError err = error("trxErr", data = "test");
```

```ballerina
error <string, map<string>> err1 = error("Error Code", message = "Msg");
(string, map<any>)|error t1 = err1;
match t1 {
    var (reason, detail) => return "Matched with tuple : " + reason + " " + "io:sprintf(\"%s\", detail)";
    var error(var reason, message = message) => return "Matched with error : " + reason + " {\"message\":\"" + <string>message + "\"}";
}
```

```ballerina
error<string, map<string|boolean>> err1 = error("Error One", message = "msgOne", fatal = true);
error<string, map<string|boolean>> err2 = error("Error Two", message = "msgTwo", fatal = false);
error<string, map<string|boolean>> err3 = error("Error Three", message = "msgThree", fatal = true);
error<string, map<string|boolean>>[3] errorArray = [err1, err2, err3];

string result1 = "";
foreach var error(reason, message = message, fatal = fatal) in errorArray {
    result1 += reason + ":";
    result1 += io:sprintf("%s:", message);
    result1 += io:sprintf("%s:", fatal);
}
```

```ballerina
type SMS error <string, map<string>>;
type SMA error <string, map<anydata>>;

function testBasicErrorVariableWithMapDetails()  {
    SMS err1 = error("Error One", message = "Msg One", detail = "Detail Msg");
    SMA err2 = error("Error Two", message = "Msg Two", fatal = true );

    string reason11;
    map<string> detail11;
    string reason12;
    string? message12;
    string? detail12;
    string? extra12;

    error (reason11, ... detail11) = err1;
    error (reason12, message = message12, detail = detail12, extra = extra12) = err1;

    string reason21;
    map<any> detail21;
    string reason22;
    any message22;
    any detail22;
    any extra22;

    error (reason21, ... detail21) = err2;
    error (reason22, message = message22, detail = detail22, extra = extra22) = err2;
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (https://github.com/ballerina-platform/ballerina-lang/issues/15535)
- [] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BConstructorSymbol.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/*
+*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BConstructorSymbol.java]: fix year

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal [hunk 2]
```diff
@@ -725,6 +728,18 @@ type InstructionGenerator object {
         self.storeToVar(typeTestIns.lhsOp.variableDcl);
     }
 
+    function generateIsLikeIns(bir:TypeTest typeTestIns) {
```
Comment (rdhananjaya) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal]: should be able to remove this

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/ErrorValue.java [hunk 3]
```diff
@@ -43,7 +45,7 @@
 
     public ErrorValue(String reason, Object details) {
         super(reason);
-        this.type = BTypes.typeError;
+        this.type = new BErrorType(TypeConstants.ERROR, null, BTypes.typeString, TypeChecker.getType(details));
```
Comment (rdhananjaya) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/ErrorValue.java]: Don't need this with new constructor (check?)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/SymTag.java [hunk 4]
```diff
@@ -50,4 +50,5 @@
     public static final int CHANNEL = 1 << 27;
     public static final int CONSTANT = 1 << 28 | VARIABLE_NAME;
     public static final int FUNCTION_TYPE = 1 << 29 | TYPE | VARIABLE_NAME;
+    public static final int CONSTRUCOR = 1 << 30 | INVOKABLE;
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/SymTag.java]: Typo. `CONSTRUCOR` -> `CONSTRUCTOR`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 5]
```diff
@@ -1353,6 +1360,14 @@ public void visit(BLangTrapExpr trapExpr) {
         }
     }
 
+    private BIROperand createErrorTargetOperandForNilReturnInvocation() {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java]: Seems like no usage.

**Pull request merged**